### PR TITLE
Add OffsetDate

### LINKED ISSUE #137: Add OffsetDate
After adding support for ThreeTen-Extra in [threeten-jaxb/threeten-jaxb](https://github.com/threeten-jaxb/threeten-jaxb) I want to tackle better compatibility with the XML schema definitions, especially [`xs:date`](https://www.w3.org/TR/xmlschema-2/#date).

You already explained, why `OffsetDate` even exists in a [Stack Overflow answer](https://stackoverflow.com/a/7841875/8400725).

I found that `OffsetDate` was removed from ThreeTen, because there are only few use cases: https://github.com/ThreeTen/threeten/issues/228

Others are reimplementing `OffsetDate` on their own, like [Addicticks/jaxbdatetime](https://github.com/Addicticks/jaxbdatetime/blob/master/src/main/java/com/addicticks/jaxb/adapters/time/OffsetDate.java), but I think, it might be better to have it in ThreeTen-Extra. What are your thoughts on this?

### PR DESCRIPTION
**Pull request by mwkroening**
Resolves #137 

Relevant ThreeTenBP commit: https://github.com/ThreeTen/threetenbp/commit/d9ad1de4c63817275c27a464781a08cb3d3d920f

#### Checklist:
- [x] Add `OffsetDate.java`
- [x] Fix `OffsetDate` compilation
  * Implemented `OffsetDate.isSupported(TemporalUnit)`
  * Removed custom serialization
- [x] Add `TestOffsetDate.java`
- [x] Fix `TestOffsetDate` compilation
- [x] Fix `TestOffsetDate` tests
  * Removed custom serialization testing
- [x] Align API with those of `OffsetDateTime` and `OffsetTime`
  - [x] Rename `toString(DateTimeFormatter)` to `format(DateTimeFormatter)`
  - [x] Rename `getDate()` to `toLocalDate()`
  - [x] Implement `from(int, int, int, ZoneOffset)`
  - [x] Implement `toEpochSecond(LocalTime)`
  - [x] Rename `withOffset(ZoneOffset)` to `withOffsetSameLocal(ZoneOffset)`
- [x] Improve test coverage
- [x] Miscellaneous
  - [x] Add Joda-Convert annotations
  - [x] `assertEquals(true, ...)` -> `assertTrue(...)` etc.
  - [x] Update copyright header
  - [x] Add contributor

### CODE REVIEW

File: src/main/java/org/threeten/extra/OffsetDate.java [hunk 1]
```diff
@@ -1278,40 +1319,15 @@ public String toString() {
...
-
-    void writeExternal(DataOutput out) throws IOException {
-        date.writeExternal(out);
-        offset.writeExternal(out);
-    }
-
-    static OffsetDate readExternal(DataInput in) throws IOException {
-        LocalDate date = LocalDate.readExternal(in);
-        ZoneOffset offset = ZoneOffset.readExternal(in);
-        return OffsetDate.of(date, offset);
```
Comment (mwkroening) [inline src/main/java/org/threeten/extra/OffsetDate.java]: I did not know what to do with these helper methods for implementing externalizable, so I removed them for now.

**Comment by mwkroening** [comment]
@jodastephen I think, this is ready for review.

I committed the original files and my changes to them separately, so it's easier to comprehend my changes.

I still have to update the copyright headers. Should I add myself as well or rather not?

**Comment by jodastephen** [comment]
Thanks for the work so far. The license should be adjusted to match that of ThreeTen-Extra exactly. Instead of adding your name there, you add it to `pom.xml` (in alphabetical order). You also add an entry in `changes.xml` using your github id.

The abstract test case should be merged into the concrete one - no point in having the abstract test case now.

The last step is to check the methods against those in `OffsetTime` and `OffsetDateTime`. Any missing methods should be added (working just from the Javadoc, not copying the JDK source code. For example, `getDate()` should be `toLocalDate()`.

On tests, the goal of the coverage check is to keep coverage at that level. If you can find the coverage gaps and add new tests, that would be great.

thanks

**Comment by mwkroening** [comment]
Thanks for your comment! I'll work on it.

I think, the `AbstractDateTimeTest` could be used to extend tests of other `Temporal`s like `YearQuarter`. Would you still like me to integrate it into `TestOffsetDate`?

**Comment by mwkroening** [comment]
I addressed your points except the aforementioned one.

I'm not 100% sure about the changes I did to the API to align it with the `java.time` one.

**Comment by mwkroening** [comment]
What can I do to make this PR progress?

**Comment by jodastephen** [comment]
Nothing at your end, I just have to find the time to review.

**Comment by mwkroening** [comment]
Sure, take your time!

**Comment by jodastephen** [comment]
All looks great, thanks a lot, and sorry for the slow response time.

**Comment by jodastephen** [comment]
Merged manually including a rebase

**Comment by mwkroening** [comment]
Thanks a lot!

**Pull request closed without merging**